pkg/pillar/cmd/zedagent/parseconfig.go

```diff
@@ -1331,55 +1331,104 @@ func parseIpspec(ipspec *zconfig.Ipspec,
-				s, err))
-		}
-		config.Subnet = *subnet
-	}
-	// Parse Gateway
-	if g := ipspec.GetGateway(); g != "" {
-		config.Gateway = net.ParseIP(g)
-		if config.Gateway == nil {
-			return errors.New(fmt.Sprintf("bad gateway IP %s",
-				g))
-		}
-	}
 	// Parse NTP Server
 	if n := ipspec.GetNtp(); n != "" {
 		config.NtpServer = net.ParseIP(n)
 		if config.NtpServer == nil {
-			return errors.New(fmt.Sprintf("bad ntp IP %s",
-				n))
+			errStr := fmt.Sprintf("bad ntp IP %s", n)
+			return errors.New(errStr)
 		}
 	}
 	// Parse Dns Servers
 	for _, dsStr := range ipspec.GetDns() {
 		ds := net.ParseIP(dsStr)
 		if ds == nil {
-			return errors.New(fmt.Sprintf("bad dns IP %s",
-				dsStr))
+			errStr := fmt.Sprintf("bad dns IP %s", dsStr)
+			return errors.New(errStr)
 		}
 		config.DnsServers = append(config.DnsServers, ds)
 	}
+	// Parse Subnet
+	if s := ipspec.GetSubnet(); s != "" {
+		_, subnet, err := net.ParseCIDR(s)
+		if err != nil {
+			errStr := fmt.Sprintf("bad subnet %s: %s", s, err)
+			return errors.New(errStr)
+		}
+		config.Subnet = *subnet
+	}
+	// Parse Gateway
+	if g := ipspec.GetGateway(); g != "" {
+		config.Gateway = net.ParseIP(g)
+		if config.Gateway == nil {
+			errStr := fmt.Sprintf(" bad gateway IP %s", g)
+			return errors.New(errStr)
+		}
+	}
 	// Parse DhcpRange
 	if dr := ipspec.GetDhcpRange(); dr != nil && dr.GetStart() != "" {
 		start := net.ParseIP(dr.GetStart())
 		if start == nil {
-			return errors.New(fmt.Sprintf("bad start IP %s",
-				dr.GetStart()))
+			errStr := fmt.Sprintf("bad start IP %s", dr.GetStart())
+			return errors.New(errStr)
 		}
 		end := net.ParseIP(dr.GetEnd())
 		if end == nil && dr.GetEnd() != "" {
-			return errors.New(fmt.Sprintf("bad end IP %s",
-				dr.GetEnd()))
+			errStr := fmt.Sprintf("bad end IP %s", dr.GetEnd())
+			log.Errorf(errStr)
+			return errors.New(errStr)
 		}
 		config.DhcpRange.Start = start
 		config.DhcpRange.End = end
 	}
+
+	// when subnet is set; gateway or, dhcp range not set,
+	// take some default
+	if config.Subnet.IP != nil {
+		base := types.GetIPSubnetBase(config.Subnet.IP, config.Gateway)
+		ipStart := 1
+		if base != 0 {
+			ipStart = base
+		}
+		dhcpRangeStart := ipStart + 1
+		// get the ip range information
+		addrCount := types.GetIPAddrCountOnSubnet(config.Subnet)
+		if addrCount == 0 {
+			errStr := fmt.Sprintf("bad subnet %s", config.Subnet.IP.String())
+			return errors.New(errStr)
+		}
+		if addrCount >= types.LargeAppCount {
+			// when gateway is less than the half point,
+			// the dhcpRange starts at the half point
+			if base < addrCount/2 {
+				dhcpRangeStart = addrCount / 2
+			}
+		}
+		dhcpRangeEnd := addrCount - 2
+
+		// the default gateway is always the first address
+		//  in the subnet
+		if config.Gateway == nil {
+			config.Gateway = types.AddToIP(config.Subnet.IP, ipStart)
+			log.Warnf("network(%s), No Gateway, setting default(%s)",
+				config.Key(), config.Gateway.String())
+		}
```

Comment:
Do we need this due to existing configurations without any Gateway? Doesn't zedcloud set such a default?

Reply:
ok